ksmt-core/src/main/kotlin/org/ksmt/expr/rewrite/simplify/KFpExprSimplifier.kt

```diff
@@ -0,0 +1,696 @@
+package org.ksmt.expr.rewrite.simplify
+
+import org.ksmt.KContext
+import org.ksmt.expr.KApp
+import org.ksmt.expr.KBitVec1Value
+import org.ksmt.expr.KBitVecValue
+import org.ksmt.expr.KBvToFpExpr
+import org.ksmt.expr.KExpr
+import org.ksmt.expr.KFp32Value
+import org.ksmt.expr.KFp64Value
+import org.ksmt.expr.KFpAbsExpr
+import org.ksmt.expr.KFpAddExpr
+import org.ksmt.expr.KFpDivExpr
+import org.ksmt.expr.KFpEqualExpr
+import org.ksmt.expr.KFpFromBvExpr
+import org.ksmt.expr.KFpFusedMulAddExpr
+import org.ksmt.expr.KFpGreaterExpr
+import org.ksmt.expr.KFpGreaterOrEqualExpr
+import org.ksmt.expr.KFpIsInfiniteExpr
+import org.ksmt.expr.KFpIsNaNExpr
+import org.ksmt.expr.KFpIsNegativeExpr
+import org.ksmt.expr.KFpIsNormalExpr
+import org.ksmt.expr.KFpIsPositiveExpr
+import org.ksmt.expr.KFpIsSubnormalExpr
+import org.ksmt.expr.KFpIsZeroExpr
+import org.ksmt.expr.KFpLessExpr
+import org.ksmt.expr.KFpLessOrEqualExpr
+import org.ksmt.expr.KFpMaxExpr
+import org.ksmt.expr.KFpMinExpr
+import org.ksmt.expr.KFpMulExpr
+import org.ksmt.expr.KFpNegationExpr
+import org.ksmt.expr.KFpRemExpr
+import org.ksmt.expr.KFpRoundToIntegralExpr
+import org.ksmt.expr.KFpRoundingMode
+import org.ksmt.expr.KFpRoundingModeExpr
+import org.ksmt.expr.KFpSqrtExpr
+import org.ksmt.expr.KFpSubExpr
+import org.ksmt.expr.KFpToBvExpr
+import org.ksmt.expr.KFpToFpExpr
+import org.ksmt.expr.KFpToIEEEBvExpr
+import org.ksmt.expr.KFpToRealExpr
+import org.ksmt.expr.KFpValue
+import org.ksmt.expr.KRealNumExpr
+import org.ksmt.expr.KRealToFpExpr
+import org.ksmt.sort.KBoolSort
+import org.ksmt.sort.KBvSort
+import org.ksmt.sort.KFpRoundingModeSort
+import org.ksmt.sort.KFpSort
+import org.ksmt.sort.KRealSort
+import org.ksmt.sort.KSort
+import org.ksmt.utils.BvUtils.bvMaxValueSigned
+import org.ksmt.utils.BvUtils.minus
+import org.ksmt.utils.BvUtils.mkBvFromBigInteger
+import org.ksmt.utils.uncheckedCast
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.math.RoundingMode
+import kotlin.math.IEEErem
+import kotlin.math.absoluteValue
+import kotlin.math.round
+import kotlin.math.sqrt
+
+@Suppress("ForbiddenComment")
+interface KFpExprSimplifier : KExprSimplifierBase {
+
+    fun <T : KFpSort> simplifyEqFp(lhs: KExpr<T>, rhs: KExpr<T>): KExpr<KBoolSort> = with(ctx) {
+        if (lhs == rhs) return trueExpr
+
+        if (lhs is KFpValue<T> && rhs is KFpValue<T>) {
+            // special cases
+            if (lhs.isNan() && rhs.isNan()) return trueExpr
```

Comment:
I believe, for fp equality it is wrong: 

![image](https://user-images.githubusercontent.com/31047452/207252452-432bfa14-5cce-4958-a6ac-1aeaec805c94.png)

Reply:
`simplifyEqFp` is a case of `(= (_ Nan ...) (_ Nan ...) )` and not `fp.eq`. Or i misunderstood you.

Reply:
You understood me correctly, but I thought it is about `fp.eq`. I think it might be important to highlight because of the difference in this behaviour and equality in Java/Kotlin